### PR TITLE
Entity transform events

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
Entity transform events

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/events/EvtEntityTransform.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	@Override
+	public boolean check(Event event) {
+		if (!(event instanceof EntityTransformEvent))
+			return false;
+		EntityTransformEvent transformEvent = (EntityTransformEvent) event;
+		if (reasons != null && reasons.check(event, reason -> transformEvent.getTransformReason().equals(reason)))
+			return false;
+		if (datas != null && !datas.check(event, data -> transformEvent.getTransformedEntities().stream().anyMatch(entity -> data.isInstance(entity))))
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/events/EvtEntityTransform.java:71]: I think this can be at most four entities so it's probably not worth the overhead to set up a stream for parallelisation, but it doesn't matter.

**Review by Moderocky — APPROVED** [review]

**Review by sovdeeth — CHANGES_REQUESTED** [review]
On top of the issues with the patterns, none of the events I tried worked in testing. I listened to the EntityTransformEvent via skript reflect to make sure, and it was called every time. None of the skript events triggered:
```
[23:47:03 INFO]: [Skript] Server Version: git-Paper-17 (MC: 1.20)
[23:47:03 INFO]: [Skript] Skript Version: 2.7.0-beta3-nightly-6356c75 (skriptlang-nightly)
[23:47:03 INFO]: [Skript] Installed Skript Addons:
[23:47:03 INFO]: [Skript]  - skript-reflect v2.4-dev1 (https://github.com/TPGamesNL/skript-reflect)
[23:47:03 INFO]: [Skript]  - SkBee v2.17.0 (https://github.com/ShaneBeee/SkBee)
[23:47:03 INFO]: [Skript] Installed dependencies:
[23:47:03 INFO]: [Skript]  - WorldGuard v7.0.9+5934e49
```
```
on lightning:
    broadcast "lightning"

on unknown:
    broadcast "unknown"

on mooshroom transform:
    broadcast "sheared"

on zombie curing:
    broadcast "cured"

on zombie, skeleton or slime transform:
    broadcast "transformed"

import:
    org.bukkit.event.entity.EntityTransformEvent

on EntityTransformEvent:
    broadcast "transformed"
    broadcast " - %event.getTransformReason()%"
```
image
image

File: src/main/resources/lang/default.lang [hunk 2]
```diff
@@ -1939,6 +1939,19 @@ quit reasons:
...
+	cured: zombie cure, zombie curing, zombie cured, cured
+	drowned: entity drowned, drowned, zombie converting to drown, zombie drown
+	frozen: entity froze, entity frozen, skeleton frozen, skeleton froze, skeleton converting to stray
+	lightning: lightning, creeper super charge, super charged creeper
+	metamorphosis: tadpole convert, tadpole converted to frog, tadpole metamorphosis, metamorphosis
+	piglin_zombified: piglin zombification, piglin zombified
+	sheared: sheared, mooshroom sheared
+	split: slime split, slime splitting, split, magma slime split, magma slime splitting
+	unknown: unknown
+	infection: infection
```
Comment (sovdeeth) [inline src/main/resources/lang/default.lang:1953]: For some reason my other review lost my comments here:
These are way too loose when combined with the event pattern:
* `on unknown` is just bad
* `on lightning` overwrites the existing lightning event
* `on sheared` is misleading, I'd think it should work on sheep given the name
* `on cured`, `on infection`, `on metamorphosis`, and `on split` are ok but very lacking in context
* `on drowned` can easily be misleading

Basically these names should be much stricter or the event's pattern needs to change. Perhaps something like `on transform due to %*transformreason%`

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/events/EvtEntityTransform.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	@Override
+	public boolean check(Event event) {
+		if (!(event instanceof EntityTransformEvent))
+			return false;
+		EntityTransformEvent transformEvent = (EntityTransformEvent) event;
+		if (reasons != null && reasons.check(event, reason -> transformEvent.getTransformReason().equals(reason)))
+			return false;
+		if (datas != null && !datas.check(event, data -> transformEvent.getTransformedEntities().stream().anyMatch(data::isInstance)))
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtEntityTransform.java:68]: This uses the future entity type, which is extremely misleading. It also means the first example ``on a zombie transforming due to curing:`` would never work. You just need to check against `transformEvent.getEntity()`. 

This should have been caught if you just used the example testing code I provided above.
Here's some more if you want to use it this time:

```
on cow transform:
    broadcast "cow transformed"
    broadcast " - reason: %event-transformreason%"
    broadcast " - entity: %event-entity%"

on mooshrom transform:
    broadcast "mooshrom transformed"
    broadcast " - reason: %event-transformreason%"
    broadcast " - entity: %event-entity%"

on entity transform:
    broadcast "entity transformed"
    broadcast " - reason: %event-transformreason%"
    broadcast " - entity: %event-entity%"
```

File: src/main/resources/lang/default.lang [hunk 4]
```diff
@@ -1953,6 +1953,19 @@ quit reasons:
 	kicked: kicked
 	timed_out: timed out
 
+# -- Transform Reasons --
+transform reasons:
+	cured: zombie cure, zombie curing, zombie cured, cured
+	drowned: entity drowned, drowned, zombie converting to drown, zombie drown
+	frozen: entity froze, entity frozen, skeleton frozen, skeleton froze, skeleton converting to stray
+	lightning: lightning, creeper super charge, super charged creeper
```
Comment (sovdeeth) [inline src/main/resources/lang/default.lang:1961]: `lightning` doesn't compare properly when the cause is in a variable, but I don't know if this is very severe.
```
on witch transform:
    broadcast event-transformreason
    if event-transformreason is lightning:
        broadcast "lightning yay"
    set {_a} to event-transformreason
    if {_a} is lightning:
        broadcast "lightning yay 2"
```
image

The other reasons should also be tested for conflicts, I haven't tested them all.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/events/EvtEntityTransform.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		reasons = (Literal<TransformReason>) args[1];
+		return true;
+	}
+
+	@Override
+	public boolean check(Event event) {
+		if (!(event instanceof EntityTransformEvent))
+			return false;
+		EntityTransformEvent transformEvent = (EntityTransformEvent) event;
+		if (reasons != null && reasons.check(event, reason -> transformEvent.getTransformReason().equals(reason)))
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtEntityTransform.java:66]: also this is just backwards. It returns false if the reason matches. I didn't catch this in the previous review because the other bug obscured it, but reasons just don't work.
```
on entity transform due to lightning:
    broadcast "entity transformed due to lightning"
    broadcast " - reason: %event-transformreason%"
    broadcast " - entity: %event-entity%"

on entity transform due to infection:
    broadcast "entity transformed due to infection"
    broadcast " - reason: %event-transformreason%"
    broadcast " - entity: %event-entity%"

on entity transform:
    broadcast "entity transformed"
    broadcast " - reason: %event-transformreason%"
    broadcast " - entity: %event-entity%"
```
image

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/events/EvtEntityTransform.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+public class EvtEntityTransform extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("Entity Transform", EvtEntityTransform.class, EntityTransformEvent.class, "(entit(y|ies)|%*-entitydatas%) transform[ing] [due to %-transformreasons%]")
+				.description("Called when an entity is about to be replaced by another entity.",
+						"Examples when it's called include; when a zombie gets cured and a villager spawns, ",
+						"An entity drowns in water like a zombie that turns to a drown, ",
+						"An entity that gets frozen in powder snow, ",
+						"A mooshrom cow that when sheared, spawns a new cow.")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/events/EvtEntityTransform.java:41]: Just some small typos/grammatical fixes
```suggestion
				.description("Called when an entity is about to be replaced by another entity.",
						"Examples when it's called include; when a zombie gets cured and a villager spawns, ",
						"an entity drowns in water like a zombie that turns into a drowned, ",
						"an entity that gets frozen in powder snow, ",
						"a mooshroom cow that when sheared, spawns a new cow.")
```

File: src/main/java/ch/njol/skript/events/EvtEntityTransform.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+				.examples("on a zombie transforming due to curing:", "on mooshroom cow transforming:", "on zombie, skeleton or slime transform:")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/events/EvtEntityTransform.java:42]: `mooshroom cow` does not appear to be valid syntax for a mooshroom
```suggestion
				.examples("on a zombie transforming due to curing:", "on mooshroom transforming:", "on zombie, skeleton or slime transform:")
```

File: src/main/resources/lang/default.lang [hunk 8]
```diff
@@ -1953,6 +1953,19 @@ quit reasons:
 	kicked: kicked
 	timed_out: timed out
 
+# -- Transform Reasons --
+transform reasons:
+	cured: zombie cure, zombie curing, zombie cured, cured
+	drowned: entity drowned, drowned, zombie converting to drown, zombie drown
```
Comment (ShaneBeee) [inline src/main/resources/lang/default.lang:1959]: The entity is called a Drowned, not a Drown
```suggestion
	drowned: entity drowned, drowned, zombie converting to drowned, zombie drown
```
The tenses here confuse me, you're using "entity drowned" (past tense) and "converting" (present tense)

**Review by ShaneBeee — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/events/EvtEntityTransform.java (hunk 6 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/events/EvtEntityTransform.java:41]: It should actually be like this
```suggestion
				.description("Called when an entity is about to be replaced by another entity.",
						"Examples when it's called include; when a zombie gets cured and a villager spawns, " +
						"an entity drowns in water like a zombie that turns to a drown, " +
						"an entity that gets frozen in powder snow, " +
						"a mooshrom cow that when sheared, spawns a new cow.")
```

File: src/main/resources/lang/default.lang (hunk 4 above)
Comment (TheLimeGlass) [reply src/main/resources/lang/default.lang:1961]: I have a pending pull request for fixing literal comparison in variables.

File: src/main/java/ch/njol/skript/events/EvtEntityTransform.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+public class EvtEntityTransform extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("Entity Transform", EvtEntityTransform.class, EntityTransformEvent.class, "(entit(y|ies)|%*-entitydatas%) transform[ing] [due to %-transformreasons%]")
+				.description("Called when an entity is about to be replaced by another entity.",
+						"Examples when it's called include; when a zombie gets cured and a villager spawns, " +
+						"an entity drowns in water like a zombie that turns to a drown, " +
+						"an entity that gets frozen in powder snow, " +
+						"a mooshrom cow that when sheared, spawns a new cow.")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/events/EvtEntityTransform.java:40]:
```suggestion
						"a mooshroom that when sheared, spawns a new cow.")
```
I think when Ayham made suggestions ... this line somehow got missed.

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
just one tiny thing left

File: src/main/resources/lang/default.lang [hunk 10]
```diff
@@ -1953,6 +1953,19 @@ quit reasons:
...
+	cured: zombie cure, zombie curing, zombie cured, cured
+	drowned: entity drowned, drowned, zombie converting to drowned, zombie drown
+	frozen: entity froze, entity frozen, skeleton frozen, skeleton froze, skeleton converting to stray
+	lightning: lightning, creeper super charge, super charged creeper
+	metamorphosis: tadpole convert, tadpole converted to frog, tadpole metamorphosis, metamorphosis
+	piglin_zombified: piglin zombification, piglin zombified
+	sheared: sheared, mooshroom sheared
+	split: slime split, slime splitting, split, magma slime split, magma slime splitting
+	unknown: unknown
+	infection: infection
```
Comment (sovdeeth) [inline src/main/resources/lang/default.lang:1967]: I think most of these should be updated to reflect the change in tense in the event syntax:
(and removing sheared, just to avoid players thinking that sheep would be involved)
```suggestion
	cured: zombie cure, zombie curing, curing
	drowned: entity drowning, drowning, zombie converting to drowned, zombie drowning
	frozen: entity freezing, entity freeze, skeleton freezing, skeleton freeze, skeleton converting to stray
	lightning: lightning, creeper super charge
	metamorphosis: tadpole converting, tadpole converting to frog, tadpole metamorphosis, metamorphosis
	piglin_zombified: piglin zombification
	sheared: mooshroom shear, mooshroom shearing
	split: slime split, slime splitting, split, magma slime split, magma slime splitting
	unknown: unknown
	infection: infection, villager infection
```
Don't forget to update examples too.

This is, of course, up to interpretation. feel free to pick what you think makes sense or if you have better ideas, let me know.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by ShaneBeee — APPROVED** [review]
